[TG-8264] Expression simplification for endsWith(), isEmpty(), and compareTo()

danpoe opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 11170eb9ec44894594fa5b567a61db74a5449088).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115279384

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -235,6 +235,50 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
   }
+  else if(func_id == ID_cprover_string_endswith_func)
+  {
+    const refined_string_exprt &s1 =
```
See the remark on the recently merged PR: it would be preferable to factor each of the simplifications into their own local/`static` procedure.

allredj replied:
Done

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -297,6 +297,78 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+        else
+        {
+          res = 1;
+        }
```
Nit pick: `res = c1 < c2 ? -1 : 1;` would condense 8 lines into a single one.

allredj replied:
Function rewritten and simplified

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -297,6 +297,78 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+
+    const array_exprt &s2_data = s2_data_opt->get();
+
+    int res;
```
Why is `res` declared here when only one branch below uses it?

allredj replied:
Refactored.

smowton commented on src/util/simplify_expr.cpp:
```diff
@@ -235,6 +235,140 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+    const refined_string_exprt &s =
+      to_string_expr(function_app.arguments().at(0));
+
+    const auto s_data_opt = try_get_string_data_array(s, ns);
```
I'd expect a null data array to be valid too, especially for a constant empty string (Java strings are counted so don't need null terminators) -- how about just checking the length field instead?

allredj replied:
Changed to checking the length

smowton commented on jbmc/regression/jbmc-strings/CompareToConstantEvaluation/Main.java:
```diff
@@ -0,0 +1,14 @@
+public class Main {
+    public void constantCompareTo() {
```
These tests should really also consider cases that *shouldn't* be simplified, such as "ab".endsWith("abc"). Try to include a case for each `if (...) return true` below.

allredj replied:
Not sure I understand. The example your mention *should* actually  be simplified, though the assertions should fail. I'm currently writing tests for failing assertions.

smowton commented on src/util/simplify_expr.cpp:
```diff
@@ -235,6 +235,140 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+    if(s1_data.operands().size() == s2_data.operands().size() &&
+       s1_data.operands() == s2_data.operands())
+    {
+      expr = from_integer(0, expr.type());
```
`res = 0`

allredj replied:
refactored

romainbrenguier commented on jbmc/regression/jbmc-strings/IsEmptyConstantEvaluation/test.desc:
```diff
@@ -0,0 +1,9 @@
+Main.class
+--function Main.constantIsEmpty
+^Generated [0-9]+ VCC\(s\), 0 remaining after simplification$
+^VERIFICATION SUCCESSFUL$
```
We should also have tests for which the verification should fail

allredj replied:
Added

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -235,6 +235,140 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+
+    bool res = false;
+
+    if(s2_data.operands().size() > s1_data.operands().size())
```
```
const bool res = s2_data.operands().size() <= s1_data.operands().size()
  && std::equal(
￼         s2_data.operands().rbegin(),
￼         s2_data.operands().rend(),
￼         s1_data.operands().rbegin());
```

allredj replied:
refactored

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -235,6 +235,140 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+      return true;
+    }
+
+    const array_exprt &s_data = s_data_opt->get();
```
used only once, can be inlined below (or use `s_data->operands()` directly)

allredj replied:
That went away I think

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -235,6 +235,140 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
+      {
+        mp_integer c1 =
+          numeric_cast_v<mp_integer>(to_constant_expr(*it_pair.first));
+        mp_integer c2 =
```
consts

allredj replied:
done

Commit:
Make try_get_string_data_array a non-member function

It doesn't need to be. It's truly static.
We want to be able to call it from local static functions.

danpoe committed:
Implement simplify_expr() for endsWith()

danpoe committed:
Implement simplify_expr() for isEmpty()

danpoe committed:
Implement simplify_expr() for compareTo()

allredj commented:
The code has been completely refactored and some tests added. I squashed all the commits because I had to rebase on the refactored simplifier anyway: https://github.com/diffblue/cbmc/pull/4830.

More exhaustive tests have been added.

danpoe commented on src/util/simplify_expr.cpp:
```diff
@@ -153,6 +158,148 @@ simplify_exprt::simplify_popcount(const popcount_exprt &expr)
   return unchanged(expr);
 }
 
+/// Simplify String.endsWith function when argument are constant
```
`argument` -> `arguments`

allredj replied:
done

danpoe commented on src/util/simplify_expr.cpp:
```diff
@@ -153,6 +158,148 @@ simplify_exprt::simplify_popcount(const popcount_exprt &expr)
+/// Simplify String.endsWith function when argument are constant
+/// \param expr [in,out]: the expression to simplify
+/// \param ns: namespace
+/// \return: true iff expr was not modified
```
The documentation should be updated to reflect changed interface using `resultt` (also for `isEmpty()` and `compareTo()`.

allredj replied:
done

danpoe commented on src/util/simplify_expr.cpp:
```diff
@@ -153,6 +158,148 @@ simplify_exprt::simplify_popcount(const popcount_exprt &expr)
+  const namespacet &ns)
+{
+  const function_application_exprt &function_app =
+    to_function_application_expr(expr);
```
Cast is now unnecessary (also for `isEmpty()` and `compareTo()`).

allredj replied:
removed

danpoe commented on src/util/simplify_expr.cpp:
```diff
@@ -153,6 +158,148 @@ simplify_exprt::simplify_popcount(const popcount_exprt &expr)
+  if(!numeric_length.has_value())
+  {
+    return simplify_exprt::unchanged(expr);
+  }
```
Could this become an invariant `INVARIANT(numeric_length.has_value(), ...)` instead?

romainbrenguier replied:
or use numeric_cast_v, there is no reason converting a constant to a mp_integer should fail

allredj replied:
followed @romainbrenguier's suggestion

danpoe reviewed (commented):
The changes look good to me. I think the functions `simplify_string_endswith_func()`, etc. should be non-static member functions of `simplify_exprt` just like all the other simplification functions (then we'd also don't need to pass `ns` and could get rid of the `simplify_exprt::` prefix for `unchanged(expr)`).

allredj commented:
The very useful code coverage tool shows that the string solver is not longer tested for `compare_to`. That's proof that constant propagation does work, but I now need to add better tests for the string solver.

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -63,6 +63,11 @@ struct simplify_expr_cachet
 simplify_expr_cachet simplify_expr_cache;
 #endif
 
+static optionalt<std::reference_wrapper<const array_exprt>>
```
:pick: optional reference_wrappers can be replaced by pointers

allredj replied:
not done

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -158,6 +158,50 @@ simplify_exprt::simplify_popcount(const popcount_exprt &expr)
+  }
+
+  const array_exprt &s1_data = s1_data_opt->get();
+
```
:pick: I find that skipping a line between each instruction makes seeing the structure more difficult

allredj replied:
compacted the code

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -232,6 +232,74 @@ static simplify_exprt::resultt<> simplify_string_is_empty(
+  const namespacet &ns)
+{
+  const function_application_exprt &function_app =
+    to_function_application_expr(expr);
```
unnecessary

allredj replied:
removed

romainbrenguier commented on src/util/simplify_expr.cpp:
```diff
@@ -232,6 +232,74 @@ static simplify_exprt::resultt<> simplify_string_is_empty(
+
+  const mp_integer c1 =
+    numeric_cast_v<mp_integer>(to_constant_expr(*it_pair.first));
+  const mp_integer c2 =
```
:pick: can you find better names than `c1`, `c2`?

allredj replied:
no, but I found longer ones: `char1` and `char2`.

romainbrenguier replied:
still slightly better

romainbrenguier reviewed (commented):
It is alright appart from the things already mentioned by @danpoe

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e6c8c7de398e842838956a3b5c64b2d3d62f191b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116653724

Commit:
Separate tests String.regionMatches

These are all skipped for now, as String.regionMatches() is not
implemented.

Commit:
Test String.compareTo more systematically and use nondet strings

Using nondet strings ensures that the string solver is invoked, and the
string operation is not replaced during constant propagation.

Commit:
Delete old String.compareTo tests

allredj commented:
Now rebased onto https://github.com/diffblue/cbmc/pull/4836 to avoid code coverage loss (only review 4 last commits)
Ready for re-review.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5ca4c1f70401ad4f84f690245d26ed4d113381bf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116809769
